ydb/library/yql/dq/opt/dq_opt_hopping.cpp

```diff
@@ -0,0 +1,794 @@
+        .Ptr();
+}
+
+TExprNode::TPtr BuildSaveHopLambda(const TCoAggregate& aggregate, TExprContext& ctx) {
+    const auto pos = aggregate.Pos();
+    const auto aggregateHandlers = aggregate.Handlers();
+
+    const auto saveStateArg = Build<TCoArgument>(ctx, pos).Name("state").Done();
+
+    TVector<TExprBase> structItems;
+    structItems.reserve(aggregateHandlers.Size());
+
+    for (const auto& handler : aggregateHandlers) {
+        const auto tuple = handler.Cast<TCoAggregateTuple>();
+        const TString columnName = BuildColumnName(tuple.ColumnName());
+
+        const auto member = Build<TCoMember>(ctx, pos)
+            .Struct(saveStateArg)
+            .Name().Build(columnName)
+            .Done();
+
+        structItems.push_back(Build<TCoNameValueTuple>(ctx, pos)
+            .Name().Build(columnName)
+            .Value<TExprApplier>()
+                .Apply(tuple.Trait().Cast<TCoAggregationTraits>().SaveHandler())
+                .With(0, member)
+                .Build()
+            .Done());
+    }
+
+    return Build<TCoLambda>(ctx, pos)
+        .Args({saveStateArg})
+        .Body<TCoAsStruct>()
+            .Add(structItems)
+            .Build()
+        .Done()
+        .Ptr();
+}
+
+TExprNode::TPtr BuildLoadHopLambda(const TCoAggregate& aggregate, TExprContext& ctx) {
+    const auto pos = aggregate.Pos();
+    const auto aggregateHandlers = aggregate.Handlers();
+
+    TCoArgument loadStateArg = Build<TCoArgument>(ctx, pos).Name("state").Done();
+
+    TVector<TExprBase> structItems;
+    structItems.reserve(aggregateHandlers.Size());
+
+    for (const auto& handler : aggregateHandlers) {
+        const auto tuple = handler.Cast<TCoAggregateTuple>();
+        const TString columnName = BuildColumnName(tuple.ColumnName());
+
+        const auto member = Build<TCoMember>(ctx, pos)
+            .Struct(loadStateArg)
+            .Name().Build(columnName)
+            .Done();
+
+        structItems.push_back(Build<TCoNameValueTuple>(ctx, pos)
+            .Name().Build(columnName)
+            .Value<TExprApplier>()
+                .Apply(tuple.Trait().Cast<TCoAggregationTraits>().LoadHandler())
+                .With(0, member)
+                .Build()
+            .Done());
+    }
+
+    return Build<TCoLambda>(ctx, pos)
+        .Args({loadStateArg})
+        .Body<TCoAsStruct>()
+            .Add(structItems)
+            .Build()
+        .Done()
+        .Ptr();
+}
+
+TMaybe<bool> BuildWatermarkMode(
+    const TCoAggregate& aggregate,
+    const TCoHoppingTraits& hoppingTraits,
+    TExprContext& ctx,
+    bool analyticsMode,
+    bool defaultWatermarksMode,
+    bool asyncActor)
+{
+    const bool enableWatermarks = !analyticsMode &&
+        defaultWatermarksMode &&
+        hoppingTraits.Version().Cast<TCoAtom>().StringValue() == "v2";
+    if (enableWatermarks && asyncActor) {
+        ctx.AddError(TIssue(ctx.GetPosition(aggregate.Pos()), "Watermarks should be used only with async compute actor"));
```

Comment:
Maybe 'Watermarks should be used only with sync compute actor' ?

Reply:
Fixed